toolbackup/src/main/java/io/dockstore/tooltester/client/cli/Client.java

```diff
@@ -0,0 +1,267 @@
+    }
+
+    private void createFileFromString(String input, String path) {
+        FileGenerator.inputStreamToFile(InputStreamGenerator.stringToInputStream(input), path);
+    }
+
+    private void report(Map<String, List<VersionDetail>> toolNameToList, String basePath) {
+        for (Map.Entry<String, List<VersionDetail>> entry : toolNameToList.entrySet()) {
+            /** Generates html page listing all the versions of the current tool */
+            createFileFromString(ReportGenerator.generateFilesReport(entry.getValue()), basePath+entry.getKey()+".html");
+        }
+        /** Generates main menu */
+        createFileFromString(ReportGenerator.generateMenu(toolNameToList.keySet(), getTotalSizeInB(toolNameToList)), basePath+"index.html");
+        /** Generates JSON map */
+        createFileFromString(ReportGenerator.generateJSONMap(toolNameToList), basePath+"JSONMap.txt");
+    }
+
+    /**
+     * Save docker images to a local directory
+     * @param basePath
+     */
+    private Map<String, List<VersionDetail>> save(String basePath) {
+        /** use swagger-generated classes to talk to dockstore */
+        setupClientEnvironment();
+
+        InputStream inputStream;
+        Map<String, List<VersionDetail>> toolNameToList = new HashMap<>();
+
+        try {
+            final DockerClient docker = DefaultDockerClient.fromEnv().build();
+            double totalSizeInBytes = 0, fileSizeInBytes;
+
+            /** removed final modifier for testing purposes only */
+            List<Tool> tools = ga4ghApi.toolsGet(null, null, null, null, null, null, null, null, null);
+
+            /** only look at 2 elements for testing purposes only */
+            final int fromIndex = 2;
+            final int toIndex = 4;
+            tools = tools.subList(fromIndex, toIndex);
+
+            String toolName, dirPath, img, versionId, versionTag, date;
+
+            final int indexAfterDate = 10;
+
+            for(Tool tool : tools)  {
+                toolName = tool.getToolname();
+                dirPath = basePath + tool.getId();
+                List<ToolVersion> versions = tool.getVersions();
+                List<VersionDetail> listOfVersions = new ArrayList<VersionDetail>();
+
+                DirectoryGenerator.createDirectory(dirPath);
+
+                for(ToolVersion version : versions) {
+                    img = version.getImage();
+                    if(img != "null") {
+                        docker.pull(img);
+                    }
+                    /** Save docker image to a .tar file and have the filename match its tag */
+                    inputStream = docker.save(img);
+                    versionId = version.getId();
+                    versionTag = versionId.substring(versionId.lastIndexOf(":") + 1);
+                    date = version.getMetaVersion().substring(0, indexAfterDate);
+                    File outputFile = FileGenerator.inputStreamToFile(inputStream, dirPath + "/" + versionTag + ".tar");
```

Comment:
This could be done more cleanly using a library like https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/FileUtils.html#copyInputStreamToFile(java.io.InputStream,%20java.io.File)